cpu/esp*: add Kconfig

gschorcht opened the pull request:
### Contribution description

This PR adds `Kconfig` for ESP32 and ESP8266 CPUs as well as the ESP netdev drivers `esp_wifi`
and `esp_now`.

### Testing procedure

Use `examples/gnrc_networking` with commands
```
USEMODULE='esp_wifi esp_now' make BOARD=esp32-wroom-32 -C examples/gnrc_networking menuconfig
USEMODULE='esp_wifi esp_now' make BOARD=esp8266-esp-12x -C examples/gnrc_networking menuconfig
```
to configure both CPUs and try to compile them afterwards.

### Issues/PRs references

Part of issue #12888

gschorcht commented:
Just force-pushed typo and whitespace fixes for static tests.

leandrolanzieri commented on cpu/esp32/esp-wifi/esp_wifi_params.h:
```diff
@@ -22,37 +22,77 @@
-#ifndef ESP_WIFI_STACKSIZE
-#define ESP_WIFI_STACKSIZE  (THREAD_STACKSIZE_DEFAULT)
+#ifndef CONFIG_ESP_WIFI_STACKSIZE
+#define CONFIG_ESP_WIFI_STACKSIZE   ESP_WIFI_STACKSIZE
```
Namespace

leandrolanzieri commented on cpu/esp32/esp-wifi/esp_wifi_params.h:
```diff
@@ -22,37 +22,77 @@
-#ifndef ESP_WIFI_PRIO
-#define ESP_WIFI_PRIO       (GNRC_NETIF_PRIO)
+#ifndef CONFIG_ESP_WIFI_PRIO
+#define CONFIG_ESP_WIFI_PRIO        ESP_WIFI_PRIO
```
Namespace

gschorcht commented on cpu/esp8266/esp-wifi/Kconfig:
```diff
@@ -0,0 +1,29 @@
+# Copyright (c) 2020 Gunar Schorcht
```
Yes, when RIOT was ported to both platforms, the ESP8266 SDK required an implementation of `esp_wifi` that was completely different from that for ESP32.

With the new ESP8266 SDK and the complete reimplementation of the RIOT port on ESP8266 in PR #11108, a new `esp_wifi` module was introduced for ESP8266 that is almost the same as for ESP32. That's why there is already PR #12955 which will do a complete code deduplication. Module `esp_wifi` will then be moved to `esp_common` completely.

But before PR #12955 is merged, we have to do it in both implementations :worried: I wished PR #12955 would be merged soon, doing everything twice is quite nerving.

leandrolanzieri replied:
But, we can just place this file in `cpu/esp_common/Kconfig.wifi` and source it from both right?

gschorcht replied:
> But, we can just place this file in `cpu/esp_common/Kconfig.wifi` and source it from both right?

Yes, this should be possible.

gschorcht commented on cpu/esp_common/esp-now/esp_now_params.h:
```diff
@@ -30,42 +88,42 @@
  */
-#ifndef ESP_NOW_STACKSIZE
-#define ESP_NOW_STACKSIZE       (THREAD_STACKSIZE_DEFAULT)
+#ifndef CONFIG_ESP_NOW_STACKSIZE
```
I decided to rename it for two reasons:

1. I wished the stack size would be configurable, but I couldn't find an usable approach til now. The option has to be an `int` which doesn't allow a default like `THREAD_STACKSIZE_DEFAULT`. Maybe, I will add a combination of a `bool` option for using the default and an `int` option for specifying a different value.

2. I wanted to avoid naming inconsistencies from the users point of view in documentations where you have tables like

   Configuration Parameter    | Description
   ------------------------------------|----------------
   #CONFIG_ESP_WIFI_SSID      | SSID of the AP to be used.
   #CONFIG_ESP_WIFI_PASS      | Passphrase used for the AP as clear text
   #CONFIG_ESP_WIFI_STACKSIZE | Stack size used for the WiFi netdev driver

   I wanted to avoid tables like the following:

   Configuration Parameter    | Description
   ------------------------------------|----------------
   #CONFIG_ESP_WIFI_SSID      | SSID of the AP to be used.
   #CONFIG_ESP_WIFI_PASS      | Passphrase used for the AP as clear text
   #ESP_WIFI_STACKSIZE | Stack size used for the WiFi netdev driver

  Therefore, I decided to name everything that is configurable by user also at command line in same way.

gschorcht replied:
I decided to rename it for two reasons:

1. I wished the stack size would be configurable, but I couldn't find an usable approach til now. The option has to be an `int` which doesn't allow a default like `THREAD_STACKSIZE_DEFAULT`. Maybe, I will add a combination of a `bool` option for using the default and an `int` option for specifying a different value.

2. I wanted to avoid naming inconsistencies from the users point of view in documentations where you have tables like

   Configuration Parameter    | Description
   ------------------------------------|----------------
   #CONFIG_ESP_WIFI_SSID      | SSID of the AP to be used.
   #CONFIG_ESP_WIFI_PASS      | Passphrase used for the AP as clear text
   #CONFIG_ESP_WIFI_STACKSIZE | Stack size used for the WiFi netdev driver

   I wanted to avoid tables like the following:

   Configuration Parameter    | Description
   ------------------------------------|----------------
   #CONFIG_ESP_WIFI_SSID      | SSID of the AP to be used.
   #CONFIG_ESP_WIFI_PASS      | Passphrase used for the AP as clear text
   #ESP_WIFI_STACKSIZE | Stack size used for the WiFi netdev driver

  Therefore, I decided to name everything that is configurable by user also at command line in same way.

gschorcht commented on cpu/esp_common/esp-now/esp_now_params.h:
```diff
@@ -81,8 +139,8 @@
-#ifndef ESP_NOW_KEY
-#define ESP_NOW_KEY             (NULL)
+#ifndef CONFIG_ESP_NOW_KEY
+#define CONFIG_ESP_NOW_KEY          ESP_NOW_KEY
```
The configuration of the key with Kconfig is a problem. The option would have to be either 0 or a 64 character array. I have no idea how to realize it with Kconfig. I renamed it to `CONFIG_` just for naming consistency.

gschorcht commented:
> Lastly, I know that these are CPU-specific configurations for the ESP, but having the WiFi and ESP-NOW options under `CPU` seems a bit off from the user perspective when using menuconfig.

What whould be a better place from your point of view?

leandrolanzieri commented:
> What whould be a better place from your point of view?

I would expect to see this maybe in `System > Networking > WiFi`.

One thing that can be done is to have generic 'WiFi' symbols (e.g. `CONFIG_WIFI_SSID`) that can be translated to esp-specific parameters. Something similar like [LoRaMAC](http://doc.riot-os.org/loramac_8h.html) has now with the keys

gschorcht commented:
> > What whould be a better place from your point of view?
> 
> I would expect to see this maybe in `System > Networking > WiFi`.

Hm :thinking: ok. Maybe there should be a further menu `Network Devices` and then `WiFi` and `ESP-NOW`?
```
`System > Networking > Network Devices > WiFi`
`System > Networking > Network Devices > ESP-NOW`
```

> One thing that can be done is to have generic 'WiFi' symbols (e.g. `CONFIG_WIFI_SSID`)

As long a there is no other WiFi netdev driver I would leave the symbols as they are. OK?

leandrolanzieri commented:
> As long a there is no other WiFi netdev driver I would leave the symbols as they are. OK?

The problem with having the ESP specific symbols is that then you need to source a file that is placed in `cpu/...` from a file that is in `sys/net/...`, which may seem odd as we are trying to follow the same structure as we have for modules.

gschorcht commented:
> > As long a there is no other WiFi netdev driver I would leave the symbols as they are. OK?
> 
> The problem with having the ESP specific symbols is that then you need to source a file that is placed in `cpu/...` from a file that is in `sys/net/...`, which may seem odd as we are trying to follow the same structure as we have for modules.

Yeah, that was the reason why I placed them in CPU section. Having `ESP-WiFi` in `Networking` and `ESP-NOW` in CPU section would be even more odd because `ESP-NOW` is nothing else than the WiFi interface that uses IEEE 802.11 Action Frames. So if we place ESPs network devices in `Networking`, the user would expect all of them there.

gschorcht commented:
> Something similar like [LoRaMAC](http://doc.riot-os.org/loramac_8h.html) has now with the keys

I couldn't find `Kconfig` for LoRaMAC. I found `sys/net/gnrc/link_layer/lorawan/Kconfig`, but in this example all symbols are also called `GNRC_LORAWAN_*`.

leandrolanzieri commented:
> I couldn't find `Kconfig` for LoRaMAC.

Sorry, I was just referring to the fact that they have generic `LORAMAC_` macros that are used both by the Semtech package and GNRC LoRaWAN.

leandrolanzieri commented on cpu/esp32/Kconfig:
```diff
@@ -0,0 +1,33 @@
+
+if KCONFIG_CPU_ESP32
+
+choice ESP32_DEFAULT_CPU_FREQ_MHZ
```
There is no need to associate a symbol to the choice here, as it is not extended somewhere else.

leandrolanzieri commented on cpu/esp8266/Kconfig:
```diff
@@ -0,0 +1,31 @@
+
+if KCONFIG_CPU_ESP8266
+
+choice ESP8266_CPU_FREQUENCY
```
Same here

leandrolanzieri commented:
> > > As long a there is no other WiFi netdev driver I would leave the symbols as they are. OK?
> > 
> > 
> > The problem with having the ESP specific symbols is that then you need to source a file that is placed in `cpu/...` from a file that is in `sys/net/...`, which may seem odd as we are trying to follow the same structure as we have for modules.
> 
> Yeah, that was the reason why I placed them in CPU section. Having `ESP-WiFi` in `Networking` and `ESP-NOW` in CPU section would be even more odd because `ESP-NOW` is nothing else than the WiFi interface that uses IEEE 802.11 Action Frames. So if we place ESPs network devices in `Networking`, the user would expect all of them there.
 
We could place ESP-NOW as well

leandrolanzieri commented:
@gschorcht do you think we could get this in shape for the feature freeze?

gschorcht commented:
I don't think so. I have changed too much in the meanwhile. For example, I provided a huge code deduplication so that `cpu/esp32/esp-wifi/esp_wifi_netdev.*` and `cpu/esp8266/esp-wifi/esp_wifi_netdev.*` were merged and moved to `esp_common`. That is, the of the WiFi netdev configuration has been moved to `esp_common`.

Furthermore, we couldn't find a final agreement on where to place the ESP-specific network device configurations (at least as I remember).

Finallly, with PR #13754, a further WiFi netdev driver will become available. Thus it's time to change the ESP-specific `ESP_WIFI_*` configurations to more general `WIFI_*` definitions. We should change it first before we expose the configuration in `Kconfig`.

leandrolanzieri commented:
> Thus it's time to change the ESP-specific `ESP_WIFI_*` configurations to more general `WIFI_*` definitions. We should change it first before we expose the configuration in `Kconfig`.

That's good, and would be more in the direction of one of my proposals above, in order to have the WiFi parameters in a more generic place.

Ok, I'll be tagging this for next release then. Thanks for the quick response!

miri64 commented:
@leandrolanzieri @gschorcht still valid?

gschorcht commented:
Because of PR #13754, it is time to change the ESP specific WiFi configuration to a more general one. This must be done first. Then we can expose the configurations via Kconfig.

fjmolinas commented:
#13754 is in now, maybe @leandrolanzieri can help rework this one now

leandrolanzieri commented:
@gschorcht any plans on moving forward with this one?

gschorcht commented:
> @gschorcht any plans on moving forward with this one?

Unfortunately not, I could not work on my RIOT contributions in the last months, because of other tasks related to the Corona online summer semester and the upcoming winter semester, which will be online again :worried: There are a number of open RIOT contributions just waiting to be completed. But I just don't have time to finish them and contribute at the moment.

fjmolinas commented:
@leandrolanzieri @gschorcht `esp` is one of the main items on the to-do list for Kconfig, is it interesting to revive this PR?

leandrolanzieri commented:
> @leandrolanzieri @gschorcht `esp` is one of the main items on the to-do list for Kconfig, is it interesting to revive this PR?

Sure, I'm currently working on the modelling of the esp modules (minus networking), adding the configurations as well would be nice.

gschorcht commented:
> Sure, I'm currently working on the modelling of the esp modules (minus networking)

@leandrolanzieri Is your modelling esp modules something different or will you provide a new PR which makes this one obsolete?

leandrolanzieri commented:
> > Sure, I'm currently working on the modelling of the esp modules (minus networking)
> 
> @leandrolanzieri Is your modelling esp modules something different or will you provide a new PR which makes this one obsolete?

It is something different, I will be providing a different PR (similar to #16929), but it does not make this one obsolete, as this adds configurations, not modules.

gschorcht commented:
> > @leandrolanzieri Is your modelling esp modules something different or will you provide a new PR which makes this one obsolete?
> 
> It is something different, I will be providing a different PR (similar to #16929), but it does not make this one obsolete, as this adds configurations, not modules.

So my first task would be to rebase this PR.

leandrolanzieri commented:
> It is something different, I will be providing a different PR

@gschorcht see #17232

gschorcht commented:
@leandrolanzieri Now were PR #17232 is merged, I would like to resume my work on the `Kconfig` for ESP network devices. I think this PR has to be reworked completely. Therefore, I would like to clarify the following questions before

1. We have now a category `Drivers -> Network Device Drivers` where there are already network device drivers. I would add ESP network devices there?
2. I would like to leave the configuration parameters like `ESP_WIFI_*` as documented to keep the compatibility with existing applications at user side. I would use the more general symbols `CONFIG_WIFI_*` in `Kconfig` and would map them as follows:
   ```c
   #ifndef ESP_WIFI_SSID
   #ifdef CONFIG_WIFI_SSID
   #define ESP_WIFI_SSID    CONFIG_WIFI_SSID
   #else
   #define ESP_WIFI_SSID    "RIOT_AP"
   #endif
   #endif
   ```
   Would that be ok?
3. How can I enable GNRC in `Kconfig` for tests? I have seen in https://github.com/RIOT-OS/RIOT/blob/85313ccc02cdb813d65f208c56d06992ad484130/sys/net/gnrc/Kconfig#L7-L8 that `Kconfig` for GNRC depends on `USEMODULE_GNRC`, but how can I enable it? According to the documentation `USEMODULE_` is generated by `Kconfig` from `USEMODULE += gnrc`. But it works neither at command line nor in Makefiles.
